WordPress/src/test/java/org/wordpress/android/ui/qrcodeauth/QRCodeAuthViewModelTest.kt

```diff
@@ -1,30 +1,553 @@
     @Before
     fun setUp() {
-        viewModel = QRCodeAuthViewModel(uiStateMapper, networkUtilsWrapper, validator)
+        viewModel = QRCodeAuthViewModel(
+                store,
+                uiStateMapper,
+                networkUtilsWrapper,
+                validator
+        )
+
+        whenever(networkUtilsWrapper.isNetworkAvailable()).thenReturn(true)
+        whenever(validator.isValidUri(scannedValue)).thenReturn(true)
+        whenever(validator.extractQueryParams(scannedValue)).thenReturn(validQueryParams)
+    }
+
+    @Test
+    fun `given empty instance state, when vm started, then loading is followed by scanning`() = runBlockingTest {
+        val result = mutableListOf<QRCodeAuthUiState>()
+        val job = launch {
+            viewModel.uiState.toList(result)
+        }
+
+        viewModel.start()
+
+        assert(result.first().loadingVisibility)
+        assertThat(result.last()).isInstanceOf(QRCodeAuthUiState.Scanning::class.java)
+
+        job.cancel()
+    }
+
+    @Test
+    fun `given non empty instance state, when vm started, then state is restored`() = runBlockingTest {
+        val result = mutableListOf<QRCodeAuthUiState>()
+        val job = launch {
+            viewModel.uiState.toList(result)
+        }
+
+        initAndStartVMForState(NO_INTERNET)
+
+        assert(result.first().loadingVisibility)
+        assertThat(result.last().type).isEqualTo(NO_INTERNET)
+
+        job.cancel()
+    }
+
+    @Test
+    fun `given onSavedInstanceState, when writeToBundle is invoked, then properties are written To Bundle`() {
+        startViewModel()
+
+        viewModel.writeToBundle(savedInstanceState)
+
+        verify(savedInstanceState).putString(any(), argThat { true })
+    }
+
+    @Test
+    fun `given validate state, when primary action is clicked, then state is authenticating`() = runBlockingTest {
+        val result = mutableListOf<QRCodeAuthUiState>()
+        val job = launch {
+            viewModel.uiState.toList(result)
+        }
+
+        initAndStartVMForState(VALIDATED)
+
+        (result.last() as Validated).primaryAction.clickAction()
+
+        assertThat(result.last().type).isEqualTo(AUTHENTICATING)
+
+        job.cancel()
+    }
+
+    @Test
+    fun `given validate state, when secondary action is clicked, then activity finished event`() = runBlockingTest {
+        val result = mutableListOf<QRCodeAuthUiState>()
+        val actionEvents = mutableListOf<QRCodeAuthActionEvent>()
+        val job1 = launch {
+            viewModel.uiState.toList(result)
+        }
+        val job2 = launch {
+            viewModel.actionEvents.collect {
+                actionEvents.add(it)
+            }
+        }
+
+        initAndStartVMForState(VALIDATED)
+
+        (result.last() as Validated).secondaryAction.clickAction()
+
+        assertThat(actionEvents.last()).isInstanceOf(QRCodeAuthActionEvent.FinishActivity::class.java)
+
+        job1.cancel()
+        job2.cancel()
+    }
```

Comment:
💡 We might want to create a wrapper function `runBlockingTestWithData` for `runBlockingTest` that `init` and `cancel` jobs for ui states and action events, something like this:

```
private fun runBlockingTestWithData(
    uiStates: MutableList<QRCodeAuthUiState> = mutableListOf(),
    actionEvents: MutableList<QRCodeAuthActionEvent> = mutableListOf(),
    testBody: suspend TestCoroutineScope.() -> Unit
) {
    runBlockingTest {
        val uiStatesJob = launch { viewModel.uiState.toList(uiStates) }
        val actionEventsJob = launch { viewModel.actionEvents.toList(actionEvents) }
        testBody()
         uiStatesJob.cancel()
         actionEventsJob.cancel()
    }
}

@Test
fun `given validate state, when secondary action is clicked, then activity finishes`() {
    val uiStates = mutableListOf<QRCodeAuthUiState>()
    val actionEvents = mutableListOf<QRCodeAuthActionEvent>()
    runBlockingTestWithData(uiStates, actionEvents) {
        initAndStartVMForState(VALIDATED)

        (uiStates.last() as Validated).secondaryAction.clickAction()

        assertThat(actionEvents.last()).isInstanceOf(QRCodeAuthActionEvent.FinishActivity::class.java)
    }
}
```

This `runBlockingTestWithData` function can be used across all functions where we're initializing and canceling the job in the test class reducing a few lines of code.

Reply:
@ashiagr - excellent idea. In commit [c962205](https://github.com/wordpress-mobile/WordPress-Android/pull/16713/commits/c962205bf5989e88e2dc05870fbc6b84c5f34636)

Reply:
Looks awesome! Thank you 🙇‍♀️